dds/DCPS/transport/tcp/TcpConnection.cpp

```diff
@@ -77,24 +76,44 @@ OpenDDS::DCPS::TcpConnection::TcpConnection(const ACE_INET_Addr& remote_address,
   , passive_setup_(false)
   , transport_during_setup_(0)
   , id_(0)
-  , reconnect_thread_(0)
+  , reconnect_task_(make_rch<TcpReconnectTask>())
 {
   DBG_ENTRY_LVL("TcpConnection","TcpConnection",6);
   this->reference_counting_policy().value(ACE_Event_Handler::Reference_Counting_Policy::ENABLED);
-
 }
+
 OpenDDS::DCPS::TcpConnection::~TcpConnection()
 {
   DBG_ENTRY_LVL("TcpConnection","~TcpConnection",6);
-  if (reconnect_thread_ &&
-    // This is for Windows, where join doesn't check if the thread is the same
-    // and the thread will hang itself if it tries to join itself.
-    !ACE_OS::thr_equal(ACE_OS::thr_self(), reconnect_thread_)
-  ) {
-    ACE_Thread_Manager::instance()->join(reconnect_thread_);
+  shutdown();
+  if (impl_) {
+    impl_->remove_reconnect_task(reconnect_task_);
   }
 }
 
+void
+OpenDDS::DCPS::TcpConnection::set_datalink(const OpenDDS::DCPS::TcpDataLink_rch& link)
+{
+  DBG_ENTRY_LVL("TcpConnection","set_datalink",6);
+
+  GuardType guard(reconnect_lock_);
+
+  // Keep a "copy" of the reference to the data link for ourselves.
+  link_ = link;
+  if (link_) {
+    impl_ = RcHandle<TcpTransport>(static_cast<TcpTransport*>(&link_->impl()), inc_count());
```

Comment:
```suggestion
    impl_ = TcpTransport_rch(static_cast<TcpTransport*>(&link_->impl()), inc_count());
```

There are some similar cases below -- use the *_rch typedef for consistency with the rest of the code